### PR TITLE
Add support for dict Input to DataFeeder, StreamingDataFeeder and allow for different TensorSignature for each of ModeKeys

### PR DESCRIPTION
**Pull request by abhitopia**
# Problem

The `Estimator.fit()` function takes as argument either
- (`x`, `y`, and `batch_size`) where `x` and `y` could be numpy arrays or iterators.
  - **PROS**
    1.   Easy to use.
    2.  Allows feeding data from arbitrary source as long as problem can be decomposed into `x` and `y`.
  - **CONS**
    1.  No provision to provide epoch
    2.  In case, `x` and `y` are arrays, the data aggregate must be available as opposed to reading on the fly ( say from database)
    3. Whether array or iterator, `x` and `y` can't be dictionaries. Most complex problems cannot be reduced to input matrix and output matrix and may require multiple input features matrices.
- `input_fn` - this is callback function which must return `features` and `target` tensors or dictionary of tensors.
  - **PROS**
    1. Allows feeding data from arbitrary source (in theory).
    2. returned features and targets can be dictionary thus allowing to solve complex problems which takes multiple inputs.
  - **CONS**
    1. Only found support for reading files using `read_batch_examples()`, `read_batch_features()`, `read_batch_record_features()`, etc.
    2. No support for passing placeholder and feed_fn to allow for arbitrary source of input data which don't require queue.
# Relevant discussions
1. https://github.com/tensorflow/tensorflow/pull/4696#issuecomment-253632403
2. http://stackoverflow.com/questions/39855375/how-to-use-streamingdatafeeder-as-contrib-learn-estimator-fits-input-fn
# Additional problem

Currently the self.feature_info (feature signature) must be same for training as well as evaluation. However, there can be cases where evaluation if done differently than training. For example, DualEncoderLSTM model where training just requires Context and Utterance, while Evaluation (Validation) requires context and multiple utterances.
# Solution

This PR basically combines the pros of using (`x`, `y`, and `batch_size`) with pros of input_fn by allowing `x`, `y` to be dictionary of multiple arrays. This also works for `x` and `y` iterators returning dictionary of streaming data. 

Also, added support for different feature/target signature for training and evaluation.

### CODE REVIEW

**Comment by abhitopia** [comment]
@martinwicke - This is updated version of #5000 where I accidentally deleted my branch. I have also resolved conflicts with the main in this PR.

**Comment by abhitopia** [comment]
Copied from #5000 where @ilblackdragon commented -

"Here is an update on current state:

we just split up SK compatibility from Estimator into separate class SKCompat. E.g. if one wants to use fit(x, y) they need to wrap their estimator with SKCompat(Estimator(...)). Otherwise preferred way will be .
there is a pending change (hopefully goes in today) to add pandas_input_fn that would allow you to est.fit(pandas_input_fn(x_df, y_series), ...) which would pass all the columns of the data frame as dict of features. It will also use local queue to feed data, which increases performance a bit compared to current data feeder solution.
Going forward we are thinking to remove data feeders, and rely on input functions similar to pandas_input_fn for other types of data. And SKCompat interface can just seamlessly wrap the passed data into this functions."

**Comment by abhitopia** [comment]
@ilblackdragon - fine if you remove the DataFeeders completely. What will be the equivalent for `StreamingDatafeeder` then? Last I checked, `input_fn` only support input from TFRecords files. 
For people like me, who have data coming from database, `StreamingDataFeeder` with dictionary support( this #5546) is a blessing. 

In any case, even you plan to introduce `SKCompat` and want to let go of DataFeeders in future, I think for now this PR could still be relevant?

@martinwicke - it's been over a month. :(

**Comment by tim-barnes** [comment]
@abhitopia +1

**Comment by martinwicke** [comment]
`input_fn` is not limited to any particular data source. You can look at the `pandas_input_fn` that @ilblackdragon mentioned, it reads from pandas Dataframes. A similar function reading from a dict would be trivial. 

We do want to remove data feeders, so relying on them may be short-lived. If you fix the conflicts, I'm happy to accept this change though. 

Sorry for the wait.

**Comment by ilblackdragon** [comment]
The idea to remove all DataFeeders - and replace them with input functions that are using [feeding_functions](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py).

See example for [pandas_input_fn](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/learn_io/pandas_io.py) - it already supports dicts passing.

What we need now to implement is `enqueue_data` to support also an iterators and write a something like `iter_input_fn`.

**Comment by abhitopia** [comment]
@martinwicke - I completely agree that we should let go of DataFeeders and have support for only for input_fn (which already supports dict). Meanwhile, the transition is still in progress, I guess it still makes sense to update DataFeeders with this PR. ( I have resolved the conflicts)

@martinwicke , @ilblackdragon  - Thanks for the pointers. I will definitely checkout `pandas_input_fn` and also see if I can contribute in writing `iter_input_fn` soon.

File: tensorflow/contrib/learn/python/learn/estimators/estimator.py
Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/estimators/estimator.py]: Can you use the ModeKey instead of a string constant?

File: tensorflow/contrib/learn/python/learn/estimators/estimator.py
Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/estimators/estimator.py]: Please use the enum instead of a string constant

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: no space after (, no space in thereof

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: no space after (

next line too

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: I think this assert is implied by the above code. Please remove it.

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: Can you fix these descriptions to be real sentences? I know you don't broke them, but they're hard to understand.

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: no space after (

Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: everywhere.

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: I don't understand this assertion, or the docstring for n_classes, or I don't understand how this works.

**Review by martinwicke — CHANGES_REQUESTED** [review]
Please make sure that you stick to 80 character per line.

**Comment by martinwicke** [comment]
Thanks for updating the PR. I left review comments. 

Please do evaluate pandas_input_fn, it is significantly easier, and should be easy to adapt into a new function dict_input_fn or similar which does exactly what you want. We'd love to have that function if you end up writing it.

We will eventually remove DataFeeder for the betterment of the word, so relying on this functionality will be brittle.

File: tensorflow/contrib/learn/python/learn/estimators/estimator.py
Comment (abhitopia) [inline tensorflow/contrib/learn/python/learn/estimators/estimator.py]: @martinwicke - I think you are looking at previous version of the file. It was already renamed after in my last commit.

File: tensorflow/contrib/learn/python/learn/estimators/estimator.py
Comment (abhitopia) [inline tensorflow/contrib/learn/python/learn/estimators/estimator.py]: thanks. will do.

File: tensorflow/contrib/learn/python/learn/estimators/estimator.py
Comment (abhitopia) [inline tensorflow/contrib/learn/python/learn/estimators/estimator.py]: @martinwicke - I am bit confused about this comment. train/eval/predict call `_check_inputs` inside their function call. If the features don't already have the signature, they are created. I don't see how "it is not allowed to pass different features to train/eval/predict". `self._features_info[mode]` and `self._labels_info[mode]` binds the signature for particular mode (namely 'train', 'eval' and 'predict'). One is free to use inputs with same signature across these modes though. May be I completely missed your point, please elaborate.

File: tensorflow/contrib/learn/python/learn/estimators/estimator.py
Comment (abhitopia) [inline tensorflow/contrib/learn/python/learn/estimators/estimator.py]: Using ModeKeys.TRAIN and ModeKeys.EVAL now.

File: tensorflow/contrib/learn/python/learn/estimators/estimator.py
Comment (abhitopia) [inline tensorflow/contrib/learn/python/learn/estimators/estimator.py]: Done.

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (abhitopia) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: done

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (abhitopia) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: done.

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (abhitopia) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: done.

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (abhitopia) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: done.

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (abhitopia) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: Done, docstring < 80 chars per line and hopefully makes more sense than before.

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (abhitopia) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: I have updated the DocString so it makes more sense, but the idea is that if `y` is dictionary, then `n_classes` can be either `None` ( in which case none of the labels in `y` are converted to one-hot) or must be dictionary with same keys as those in `y`.  Hope it makes more sense now?

**Comment by abhitopia** [comment]
@martin -  I have updated with changes you asked. And left comment on where I have question from you. Please review. Thanks.

**Comment by abhitopia** [comment]
@martinwicke , @ilblackdragon - Once this is accepted, I will try and work on proposed `dict/iter_input_fn` and also update my private code to use same.

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: I am very unhappy about these dict parameters that have to be in sync. It's just asking for trouble. It reinforces my desire to do away with it all.

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: nit: "(or iterable which returns dict)"

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: Can you put in backticks every expression that's code or pseudocode? See the documentation howto. The docstrings will be rendered as markdown, so in this case, say `n_classes[key] = n_classes for y[key]`, not n_classes[key] = n_classes for y[key].

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: `y`, `dict`, `n_classes`, `n_ classes[key]`, `y[key]`, `None`

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (martinwicke) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: `n_classes`

**Review by martinwicke — CHANGES_REQUESTED** [review]
Only small stuff left: mainly docstring formatting. Please enclose all code and pseudocode in backticks.

**Comment by martinwicke** [comment]
Jenkins, test this please.

File: tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py
Comment (abhitopia) [inline tensorflow/contrib/learn/python/learn/learn_io/data_feeder.py]: And I second your opinion as I mentioned before. I guess none of this will be required once we switch to `iter_input_fn`

**Comment by abhitopia** [comment]
Jenkins, test this please.

**Comment by abhitopia** [comment]
Jenkins, test this please.

**Comment by abhitopia** [comment]
@martinwicke, @ilblackdragon  - I have fixed all the bugs due to which some tests were failing. I also want to **emphasise** that this PR also adds support for different input signatures for `ModeKeys.TRAIN` and `ModeKeys.EVAL` and nicely extends it to `ModeKeys.INFER`.

As such, even if `DataFeeders` are going away ( and rightly so),  this PR still adds necessary support which makes already awesome `contrib.learn` even better.

Can you please review this soon? Many thanks.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
Bunch of things to fix.

I don't think we should allow letting the input signatures vary by mode. It seems to me this makes it more fragile -- is there a use case you have in mind for which this is useful?

@ilblackdragon do you have an opinion on this?

**Comment by abhitopia** [comment]
@martinwicke, @ilblackdragon  - I can give you at least one real life example where this is required. 
Consider the case of Response Selection in conversation.  ( See http://www.wildml.com/2016/07/deep-learning-for-chatbots-2-retrieval-based-model-tensorflow/)

**During training:**
X: context
y: response

**During evaluation:**
X: context
y: k responses which are ranked against X to give k@recall metric.

Also, I don't understand how having different signature makes it fragile. If for instance, the tensor signature in ModeKeys.EVAL is not compatible with computation graph, the error will be generated any how. Similarly, for ModeKeys.INFER the code first tries to match it up against signatures for ModeKeys.TRAIN/EVAL. If it doesn't match, warning is generated. But if the user doesn't pay heed to the warning and continues, that error should automatically get generated down the line if signature of ModeKeys.INFER is incompatible with computation graph.

**Comment by abhitopia** [comment]
@martinwicke - I am still waiting for your response to my previous comment, meanwhile, I just ran the tests on docker container using 
`tensorflow/tools/ci_build/ci_build.sh CPU bazel test //tensorflow/...`

and seems that 8 tests TIMEOUT, and no tests failed. I am not sure if that is the reason for some checks not passing. Any ideas on how to fix TIMEOUT tests?

**Comment by martinwicke** [comment]
The sanity checks failed, can you fix these errors? It may be that the other tests are stale.

**Comment by abhitopia** [comment]
@martinwicke, @ilblackdragon - struggling with running ci_build on mac #5759 . Any advise?

**Comment by abhitopia** [comment]
@martinwicke, @ilblackdragon, @caisq  - I need help here. There were no errors but only two TIMEOUT errors when i run

`tensorflow/tools/ci_build/ci_build.sh CPU bazel test //tensorflow/...` 

screen shot 2016-11-22 at 19 53 11

But I am absolutely unable to figure out why sanity checks are failing only for `datafeeder.py`.  The complain seems to be about Bad indentation. But when I compare for instance other file like `graph_io.py`, indentation looks same to me. Further more, if I run `pylint` with `ci_build/pep8` configuration file, then the same indentation errors appear for pretty much all other files in Tensorflow (which I did not modify).

screen shot 2016-11-22 at 20 12 01

I really need some input from you.

**Comment by pyrytakala** [comment]
@martinwicke, @abhitopia whats the status of this?

**Comment by caisq** [comment]
@abhitopia Please do not worry about testing it yourself. We will run test suites on our Jenkins to test this PR before merging it. Flaky timeouts like the ones you see do occur and in this case they seem to be unrelated to your changes.

**Comment by caisq** [comment]
@tensorflow-jenkins test this please

**Comment by caisq** [comment]
@abhitopia It looks like that the sanity check passed on Jenkins: https://ci.tensorflow.org/job/tensorflow-pull-requests-sanity/2181/console

**Comment by abhitopia** [comment]
@caisq , @martinwicke - The only tests that failed were due to Python3 compatibility. I have fixed them now. Please check again.

**Comment by abhitopia** [comment]
@tensorflow-jenkins test this please

**Comment by abhitopia** [comment]
@caisq - Can you test this again please?

**Comment by caisq** [comment]
@tensorflow-jenkins test this please.

**Comment by abhitopia** [comment]
@caisq - Thanks for testing. It seems the error testing is a bit flaky as now one Mac OS test is failing.
`//tensorflow/python/kernel_tests:cwise_ops_test                          FAILED in 1 out of 50 in 28.8s`

This test passed before, and now without any change, failed. This test is not related to this PR, 
can we get it merged please? 

CC: @martinwicke

**Comment by abhitopia** [comment]
@martinwicke  - any comments please?

**Comment by gunan** [comment]
MacOS tests were broken at head. I just pushed a fix for them.
Jenkins, test his please.

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by abhitopia** [comment]
@martinwicke , @gunan, @ilblackdragon, @caisq - Tests passing now. 😄  Merge?

**Comment by gunan** [comment]
All the team is out on thanksgiving, and I am not very familiar with the code.
I would feel more comfortable once we have approval from Martin.

**Comment by abhitopia** [comment]
@gunan - thanks for letting me know. I agree, let's wait for @martinwicke's approval.

**Comment by abhitopia** [comment]
@martinwicke - Can you check this please?

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
Thank you, and sorry this took so long.

**Comment by abhitopia** [comment]
@martinwicke - Thank you.

**Comment by abhitopia** [comment]
@martinwicke, @davidsoergel  - When I submitted, all the tests pass, and now you have reverted the changes.  :(

**Comment by martinwicke** [comment]
Yes, I am sorry.​

All the tests passed but it turns out the tests were not comprehensive
enough. We should try this again. Can you make another PR with exactly the
same changes, and assign me?

**Pull request merged**
